### DISCUSSION TITLE
How to render dynamic Markdoc pages in Next.js?

### DISCUSSION THREAD

**Discussion by john-tipper** [markdoc/markdoc#181] [Q&A] [answered]
Awesome tool, thank you so much. I'm currently exploring the tool with Next.js and am wondering on what the best pattern is for rendering pages dynamically, particularly using Next.js?  I'd prefer to be able to separate out my content from my presentation style without having to do a static build.

Thus, imagine there is a backend somewhere containing Markdoc files and these files are being added to: `example1.md`, `example2.md` etc, together with associated resources such as images (`example1_img.png`, `example2_img.png` etc).

What's the best approach to rendering these using Next.js? Is it to use `Markdoc.parse()`, `Markdoc.transform()` then something like `renderers.react()`, or is there a better approach?

If it's the former then I'm still a little confused by the documentation around `Markdoc.transform()` and `renderers.react()` and specifically their arguments (https://markdoc.dev/docs/render#react). Is there an example somewhere showing what the parsing of a complete Markdoc document (including front matter) into a React component looks like please?

**Comment by john-tipper** [comment, marked as answer]
Ah, so it looks like I need to do two parses to separate out the front matter from the document content, then do arbitrary user-defined logic according to what's in the front matter: https://github.com/markdoc/docs/blob/main/examples/react-nodejs/createContentManifest.js#L20-L21

**Reply by mfix-stripe** [reply]
@john-tipper you are right, you need to parse the frontmatter separately from the rest of the document. This is because Markdoc is agnostic as to what you put in the frontmatter (https://markdoc.dev/docs/frontmatter for more details). 

To answer your question though, the phases to render a document are outlined on this page: https://markdoc.dev/docs/render, namely:
1. Parse
2. Transform
3. Render

For your use case, you could do something like this in your Next.js App:

```js
import React from 'react';
import Markdoc, { renderers } from '@markdoc/markdoc';
import yaml from 'js-yaml';

const schema = {
  /* ... */
};
const components = {
  /* ... */
};

export const getStaticProps = async () => {
  const doc = `# Markdoc document`;

  const ast = Markdoc.parse(doc);

  /**
   * transform must be called in getStaticProps to support server-side rendering while
   * accessing variables on the server
   */
  const content = await Markdoc.transform(ast, {
    ...schema,
    variables: {
      /* ... */
    }
  });

  /**
   * Like the AST, frontmatter won't change at runtime, so it is loaded at file root.
   * This unblocks future features, such a per-page dataFetchingFunction.
   */
  const frontmatter = ast.attributes.frontmatter
    ? yaml.load(ast.attributes.frontmatter)
    : {};

  return {
    // Removes undefined
    props: JSON.parse(
      JSON.stringify({
        markdoc: {
          content,
          frontmatter
        }
      })
    )
  };
};

export default function Page({ markdoc: { content, frontmatter } }) {
  return renderers.react(content, React, { components });
}
```

**Reply by john-tipper** [reply]
Super helpful, thank you very much.
